pims/base_frames.py

```diff
@@ -607,3 +608,108 @@ def __repr__(self):
             s += "Axis '{0}' size: {1}\n".format(dim, self._sizes[dim])
         s += """Pixel Datatype: {dtype}""".format(dtype=self.pixel_type)
         return s
+
+
+def guess_axes(image):
+    shape = image.shape
+    ndim = len(shape)
+
+    if ndim == 2:
+        return 'yx'
+    elif ndim == 3 and shape[2] in [3, 4]:
+        return 'yxc'
+    elif ndim == 3:
+        return 'zyx'
+    elif ndim == 4 and shape[3] in [3, 4]:
+        return 'zyxc'
+    else:
+        raise ValueError("Cannot interpret dimensions for an image of "
+                         "shape {0}".format(shape))
+
+
+def default_axes(sizes, mode):
+    if mode == 'i':  # single image
+        if sizes.get('z', 1) == 1 and sizes.get('t', 1) == 1:
+            bundle_axes = 'yx'
+            iter_axes = ''
+        else:
+            raise ValueError("This file cannot be opened as single image.")
+    elif mode == 'I':  # multiple images
+        if sizes.get('z', 1) == 1 and 't' in sizes:
+            bundle_axes = 'yx'
+            iter_axes = 't'
+        else:
+            raise ValueError("This file cannot be opened as multiple images.")
+    elif mode == 'v':  # single volume
+        if 'z' in sizes and sizes.get('t', 1) == 1:
+            bundle_axes = 'zyx'
+            iter_axes = ''
+        else:
+            raise ValueError("This file cannot be opened as single volume.")
+    elif mode == 'v':  # single volume
+        if 'z' in sizes and 't' in sizes:
+            bundle_axes = 'zyx'
+            iter_axes = 't'
+        else:
+            raise ValueError("This file cannot be opened as multiple volumes.")
+    else:
+        if 'z' in sizes:
+            bundle_axes = 'zyx'
+        else:
+            bundle_axes = 'yx'
+        if 't' in sizes:
+            iter_axes = 't'
+        else:
+            iter_axes = ''
+
+    if 'c' in sizes:
+        bundle_axes += 'c'
+
+    return bundle_axes, iter_axes
+
+
+def wrap_get_data(get_data_func, axis='t'):
+    # takes only kwargs (one index per named axis)
+    def get_frame(**kwargs):
+        index = kwargs[axis]
+        im, md = get_data_func(index)
+        return Frame(im, frame_no=index, metadata=md)
+    return get_frame
+
+
+class WrapImageIOReader(FramesSequenceND):
+    def __init__(self, imageio_reader):
+        if not isinstance(imageio_reader, Format.Reader):
+            raise ValueError("Can only wrap ImageIO readers")
+
+        super(WrapImageIOReader, self).__init__()
+        self.rdr = imageio_reader
+
+        # TODO pass the dimension-awareness through this field
+        try:
+            info = self.rdr._get_pims_info()
+            if not isinstance(info, dict):
```

Comment:
I'm not sure I follow this. Why wouldn't info be a `dict`?

Reply:
Not sure what I was thinking, I will remove it when implementing the wrapping of readers that are already ND